utils/src/formatting_tools.c

```diff
@@ -316,3 +316,78 @@ int concat(uint8_t ** dest, size_t * dest_length, uint8_t * input,
   *dest_length = new_dest_len;
   return (0);
 }
+
+//############################################################################
+// convert_string_to_digest()
+//############################################################################
+int convert_string_to_digest(char *str, TPM2B_DIGEST * digest)
+{
+  // substring to holds 2 hex values at a time
+  char substr[3];
+
+  substr[2] = '\0';
+
+  size_t strlength = strlen(str);
+
+  if (strlength != (size_t) POLICY_DIGEST_HEXSTRING_SIZE)
+  {
+    return 1;
+  }
+  // initializes buffer with all 0 hex values
+  size_t digest_size = (8 * sizeof(digest)) + 2; // sizeof(digest) j
+  unsigned long ul;
+
+  unsigned char *expectedPolicyBuffer = (unsigned char *) malloc( 2*digest_size + 1 );
+  if( expectedPolicyBuffer == NULL ) return 1;
+  expectedPolicyBuffer[0] = 0x00; // patch - compiler won't do above assignment
+
+  // iterates through each pair of hex values and fills the
+  //  buffer with values indicated in the string
+  for (size_t i = 0; i < strlength; i += 2)
+  {
+    strncpy(substr, &str[i], 2);
+    ul = strtoul(substr, NULL, 16);
+    expectedPolicyBuffer[i / 2] = ul;
+  }
+
+  // converts the byte array into a TPM2B_DIGEST struct
+  memcpy(digest, expectedPolicyBuffer, sizeof(*expectedPolicyBuffer));
+  return 0;
+}
+
+//############################################################################
+// convert_digest_to_string()
+//############################################################################
+int convert_digest_to_string(TPM2B_DIGEST * digest, char *string_buf)
+{
+  // total number of hex values in the TPM2B digest
+  size_t digest_size = (8 * sizeof(digest)) + 2;
+
+  char * hex_buf;
+
+  hex_buf = (char *) malloc( digest_size+1 );
+  if( hex_buf == NULL ) return 1;
+
+  memcpy(hex_buf, digest, digest_size);
```

Comment:
Since the variable 'digest' points to a TPM2B_DIGEST struct, I think that we should be using the formatting provided by the struct to parse out only the SHA-256 digest value.

Again, I do not think that we need to create new variables 'hex_buf' and 'digest_size', when the values of the hash size and hash result are already directly available as digest->size and digest->buffer respectively.

In the for loop that produces the hex string, I think we could iterate over the hash value as:
```
for (int i = 0; i < digest->size; i++)
{
...
}
```
and access hash bytes as

```
digest->buffer[i]
```

This would also change the behavior of the code as the size bytes would no longer be included in the hex string, because we would not be computing it over all of the bytes in the struct. In my opinion, not including the length bytes is clearer and negates the need to consider endian-ness.

If we do want a new buffer, I think the memcpy should be revised to something like

```
memcpy(hex_buf, digest->buffer, digest->size);
```